src/include/migraphx/instruction.hpp

```diff
@@ -94,6 +111,9 @@ struct instruction
     std::vector<instruction_ref> output;
     std::vector<instruction_ref> arguments;
     literal lit;
+    int stream        = -1;
```

Comment:
The stream is set by `set_stream` so why does this store a stream number as well?

Reply:
Because in  "dom_info", we need to find out which stream an instruction belongs to.
If this stream info is not in the instruction, I have to traverse instruction stream until "set_stream" is found.    This is expensive.  To make it cheaper, I once inserted "set_stream" for every instruction so that the traverse will be cheaper and added another pass to remove redundant "set_stream"s.  This approach is not clean as well.  It assumes that after "pre_scheduling" pass and before the redundancy removal pass,  every instruction that uses stream must have a preceding "set_stream".   This assumption is quickly broken when I write unit tests to test memory coloring but forgot to add "set_stream" for every instruction.  This experience makes me feel that only using "set_stream" is an inferior solution.

In addition,  in the case that events are not represented as instructions,  in program::eval, we need to iterate through inputs of an instruction to find out whether "wait_event" needs to be inserted.  Here we need to know stream info of input instructions.  If stream is not inside instruction, we have to start the instruction stream traversal or have to build a side data structure to store the stream info.   Neither approach is a good idea.

Reply:
I rather not store this in the instruction class, as it could become stale and incorrect, and is also very target-specific. We can just have a function to create a mapping from instruction to stream.

```cpp
std::unordered_map<instruction_ref, int> get_instruction_streams(program & p);
```

In the future we can have the pass manager cache these results and only recompute it when necessary.

Reply:
Actually, thinking about this more. If we always add a `set_stream` instruction before each instruction, then we can get the stream by simply checking the previous instruction:

```cpp
int get_stream(instruction_ref ins, int depth=100)
{
    assert(depth > 0);
    auto prev = std::prev(ins);
    if (prev->name() == "set_stream")
        return any_cast<set_stream>(ins->get_operator()).stream;
    return get_stream(prev, depth-1);
}
```

Ideally, it should only be the previous instruction in most cases. Although, it can be possible we could have inserted one or two new instructions. Either case, the traversal should not be that deep.

We can then at the end remove the redundant `set_stream` instructions with this:

```cpp
void eliminate_set_streams(program& p)
{
    int stream = -1;
    for(auto ins:iterate_for(p))
    {
        if (ins->name() != "set_stream")
            continue;
        auto&& op = any_cast<set_stream>(ins->get_operator());
        if (op.stream == stream)
            p.replace_instruction(ins, op::undefined{}); // DCE will then remove the instruction
        stream = op.stream;
    }
}
```

Reply:
at program::eval, we need to get stream for instructions.  Do you want to call get_instruction_steams before the codes in eval is executed?   Where do we place this call?   

 Stream and events are set at the end of compilation.  Earlier pass is free to change the fields.
After compilation, program is not supposed to be modified.

Reply:
> at program::eval, we need to get stream for instructions. Do you want to call get_instruction_steams before the codes in eval is executed? Where do we place this call?

Why would we need to call it during eval? The `set_stream` operator will be called to set the current stream for the instructions.

Reply:
program.cpp shouldn't need to check the stream. `dom_info` does but we can pass in a function to retrieve the stream number.

Reply:
program.cpp needs to check the stream since inserting set_stream and record_event, wait_event slows down perf number (Mev's data shows overall it doesn't perturb a lot, but he also has one data that perturbs 4%, similar to what I observed).   I have explained that I have tried inserting set_stream for every instruction and found that it is an inferior solution.

Reply:
I have done my duty to implement your ideas and found that it is not good.  I am not going to keep on trying whatever new ideas you keeps on coming up.